Fix for mempool not clearing

SWvheerden opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Description
<!--- Describe your changes in detail -->
This gives the statemachine access to the mempool to give synced blocks to it. 

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
Currently if the statemachine syncs blocks, the mempool is not given the blocks to clear its pool. The mempool only listens for the addblock event to trigger its "cleansing". But because there is a bug in block propagation, nodes sync rather than get new blocks via propagation. This causes that the node's mempool is never cleared of all transactions

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
* [x] Bug fix (non-breaking change which fixes an issue)
* [ ] New feature (non-breaking change which adds functionality)
* [ ] Breaking change (fix or feature that would cause existing functionality to change)
* [ ] Feature refactor (No new feature or functional changes, but performance or technical debt improvements)
* [ ] New Tests
* [ ] Documentation

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
* [x] I'm merging against the `development` branch.
* [x] I ran `cargo-fmt --all` before pushing.
* [x] I have squashed my commits into a single commit.
* [ ] My change requires a change to the documentation.
* [ ] I have updated the documentation accordingly.
* [ ] I have added tests to cover my changes.

CjS77 commented:
Well done for finding the underlying cause.

However, the Base node state machine should not be coupled to the mempool. It should be possible to run a non-mining base node without a mempool; and this coupling would prevent that. Another sign that the approach is a not quite right is that the bulk of the PR is making a test much more bulky to add in a mempool.

I would recommend  rewriting the PR to use the PubSub mechanisms we have available. They were developed precisely for this purpose -- of keeping things decoupled and improving testability.

Specifically, the Base node has the ability to publish `BlockEvents` onto an event stream ([source](base_node/comms_interface/inbound_handlers.rs:276)). 

At moment this event is only created when a broadcast block is received, but can easily be updated to also publish the event during synchronisation.

The mempool can then subscribe to such messages and call `self.process_published_block(block)` accordingly.

This approach keeps major components decoupled, simplifies testing and makes use of the messaging systems as intended.

SWvheerden committed:
Use local node interface to submit blocks so that channel subscribers can read updated block chain backend
